### ISSUE TITLE
Decompostition of KernelResources

### ISSUE THREAD

**Issue by phil-levis** [tock/tock#3212]
Tock's current approach for configuring the kernel is the [`kernel::platform::platform::KernelResources`](https://github.com/tock/tock/blob/2685c145d660c885598e8e83f6f460bcafac1015/kernel/src/platform/platform.rs#L17) trait. An implementation of this trait (often by a board structure) is passed to [`kernel::kernel_loop_operation`](https://github.com/tock/tock/blob/2685c145d660c885598e8e83f6f460bcafac1015/kernel/src/kernel.rs#L438).

KernelResources currently defines seven implementations of kernel policies:
  1. SyscallDriverLookup,
  2. SyscallFilter,
  3. ProcessFault,
  4. ContextSwitchCallback,
  5. Scheduler,
  6. SchedulerTimer, and
  7. WatchDog.

[AppID will likely add an eighth, the policy for checking the integrity and authenticity of userspace binaries.](https://github.com/tock/tock/pull/2809) 

These eight traits cover a wide range of functionality, ranging from mapping kernel dependencies to hardware implementations (SchedulerTimer, Watchdog), kernel driver configuration (SyscallDriverLookup), hooks for kernel features (ContextSwitchCallback), kernel execution policies (Scheduler, ProcessFault), and security (SyscallFilter, AppID).

As this is now a flat structure, needing to change one of them means code must be able to change all of them. For example, if you want a developer to be able to reconfigure system call driver lookup (the set of drivers), they can also reconfigure the security policies. 

Decomposing this into different classes of resources could make this management easier going forward. The exact decomposition, however, is uncertain. One strawman solution might be:
  - Hardware resource mappings: Watchdog, SchedulerTimer
  - Execution policies: Scheduler, ProcessFault, ContextSwitchCallback
  - Security policies: AppID/Process credentials checking, System call filtering
  - Drivers/userspace: SyscallDriverLookup

I imagine those at the top changing the least often, with dynamisms increasing as you go down the list.

The purpose of this issue is to discuss what a more granular decomposition of KernelResources might look like and under what circumstances we might want to adopt it (the answer might be not any time soon).

**Comment by hudson-ayers** [comment]
> As this is now a flat structure, needing to change one of them means code must be able to change all of them. For example, if you want a developer to be able to reconfigure system call driver lookup (the set of drivers), they can also reconfigure the security policies.

I disagree with this statement, instead I just think that it being a flat structure makes that the path of least resistance.

https://github.com/hudson-ayers/tock/tree/kernel-resources-less provides an example of how it is possible to use the trait as it is today without oversharing unneeded portions of the resources struct.
